### PR TITLE
Partition refactoring

### PR DESCRIPTION
**Pull request by peterlimg**
## Fixes
- #1142 

## Changes

TLDR:
- Make it easy to support new partitions
- Clean APIs
- Move partition data types to their own packages
- Fixed the `GetRandomSlice()` bug
- Performance improvements
- Check [README](https://github.com/0chain/0chain/tree/staging/code/go/0chain.net/smartcontract/partitions#readme) for the details about how to use the partition package.

### Purpose
The purpose of this PR was to make the partition easy to use and easy to support new partitions, check issue #1142 for details. 

After refactoring, the `partitions` package is a pure utility package, hence no business data structs defined there. 

Main APIs are:

```go
// Create partitions if not exists
func CreateIfNotExists(state state.StateContextI, name string, size int) (*Partitions, error)

// Get partitions of given name
func GetPartitions(state state.StateContextI, name string) (*Partitions, error)

// Get random items from given partitions, the items will be returned through the last parameter
func GetRandomItems(state state.StateContextI, name string, v interface{}) error

// Get item from given partition index and item id
func (p *Partitions) GetItem(state state.StateContextI, partIndex int, id string, v interface{}) error

// Update item in given partition
func (p *Partitions) UpdateItem(state state.StateContextI, partIndex int, item PartitionItem) error

// Save saves the partitions to MPT
func (p *Partitions) Save(state state.StateContextI) error
...
```

### Unexpected performance gain
Apart from the main purpose for making the package easy to use, the performance was also improved especially the `GetItem()` method. Previously if we want to get an item from partition, we will need to decode all the items in the partition and go through the item list to find it. But now, we don't need to decode every pieces of struct fields of each item, instead we only decode the `ID` field as an index, and the remaining data are kept as `[]byte`, so only when we find the `ID` is matched, we will decode the `[]byte` into concrete type. The new `GetItem()` methods is about 20~30x faster than before if the previous way code is written correctly(i.e, use interface assert instead calling `Decode(data.Encode())` to get item), or it would be even faster.
  -  The `GetRandomItems()` (previously as `GetRandomSlice()`) is slightly faster, but since we are using reflection partially to make this API clean, I didn't expect it could be faster than the previous version, but yeah the truths is, it is faster indeed.
  - The `UpdateItem()` is also faster than before apparently, as finding the item is faster.
  - I believe remove item would be faster as well, because we don't need to decode the full item list to find the item to remove. 

## Tests 
Tasks to complete before merging PR:
- [ ]  Ensure system tests are passing. If not [Run them manually](https://github.com/0chain/0chain/actions/workflows/system_tests.yml) to check for any regressions :clipboard:
- [ ]  Do any new system tests need added to test this change? do any existing system tests need updated? If so create a PR at [0chain/system_test](https://github.com/0chain/system_test)
- [ ]  Merge your system tests PR to master AFTER merging this PR

## Associated PRs (Link as appropriate):
No

### CODE REVIEW

**Review by dabasov — APPROVED** [review]
lgtm

File: code/go/0chain.net/smartcontract/partitions/partitions.go [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+
+	"0chain.net/core/util"
+
+	"0chain.net/chaincore/chain/state"
+)
+
+// APIs
+
+type Partitions struct {
+	rp *randomSelector
```
Comment (ssardana96) [inline code/go/0chain.net/smartcontract/partitions/partitions.go]: would rs be a better name here?

File: code/go/0chain.net/smartcontract/partitions/partitions.go [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+
+type PartitionItem interface {
+	util.MPTSerializableSize
+	GetID() string
+}
+
+// CreateIfNotExists creates a partition if not exist
+// It's a common patten to call this to create partitions on
+// top-level partitions when project start
+func CreateIfNotExists(state state.StateContextI, name string, partitionSize int) (*Partitions, error) {
```
Comment (ssardana96) [inline code/go/0chain.net/smartcontract/partitions/partitions.go:25]: it can call GetPartitions

File: code/go/0chain.net/smartcontract/partitions/partitions.go [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+// GetRandomItems returns items of partition size number from random partition,
+// if the last partition is not full, it will try to get and fill it with its partition
+// of index - 1.
+func (p *Partitions) GetRandomItems(state state.StateContextI, r *rand.Rand, v interface{}) error {
+	return p.rp.GetRandomItems(state, r, v)
+}
+
+type ChangePartitionCallback = func(string, []byte, int, int, state.StateContextI) error
+
+// Partitioner interface wraps the methods for processing a partition
```
Comment (ssardana96) [inline code/go/0chain.net/smartcontract/partitions/partitions.go]: we can remove this

File: code/go/0chain.net/smartcontract/partitions/random.go [hunk 4]
```diff
@@ -67,34 +53,48 @@ func (rs *randomSelector) SetCallback(callback ChangePartitionCallback) {
...
-	item PartitionItem,
-	balances state.StateContextI,
-) (int, error) {
-	var part PartitionItemList
+func (rs *randomSelector) Add(state state.StateContextI, item PartitionItem) (int, error) {
+	var part *partition
 	var err error
 	if len(rs.Partitions) > 0 {
-		part, err = rs.getPartition(len(rs.Partitions)-1, balances)
+		part, err = rs.getPartition(state, len(rs.Partitions)-1)
```
Comment (ssardana96) [inline code/go/0chain.net/smartcontract/partitions/random.go:60]: state looks like a better name than balances.
I guess we should change everywhere to state

File: code/go/0chain.net/smartcontract/storagesc/blobber_reward_partition.go [hunk 5]
```diff
@@ -0,0 +1,49 @@
+package storagesc
+
+import (
+	"strconv"
+	"strings"
+
+	c_state "0chain.net/chaincore/chain/state"
```
Comment (ssardana96) [inline code/go/0chain.net/smartcontract/storagesc/blobber_reward_partition.go:7]: we should think of better name to distinguish them?
can chaincore/chain/state be moved to chaincore/state only?

**Review by ssardana96 — CHANGES_REQUESTED** [review]

File: code/go/0chain.net/smartcontract/partitions/partitions.go (hunk 2 above)
Comment (peterlimg) [reply code/go/0chain.net/smartcontract/partitions/partitions.go:25]: We have `GetPartitions()` function, and `Get*` function should not have `update` action

File: code/go/0chain.net/smartcontract/partitions/random.go (hunk 4 above)
Comment (peterlimg) [reply code/go/0chain.net/smartcontract/partitions/random.go:60]: Yes, but the name is conflict with the `state` package, so perhaps we need another better name. Anyway, I'll keep it here and think about it later.

File: code/go/0chain.net/smartcontract/storagesc/blobber_reward_partition.go (hunk 5 above)
Comment (peterlimg) [reply code/go/0chain.net/smartcontract/storagesc/blobber_reward_partition.go:7]: They are two difference thing, but yes, we need to think of a better package name. We can do it later perhaps

**Review by ssardana08 — APPROVED** [review]
